### PR TITLE
split requests by id

### PR DESCRIPTION
**Pull request by gregdhill**
Signed-off-by: Gregory Hill <gregorydhill@outlook.com>

### CODE REVIEW

File: contracts/Option.sol [hunk 1]
```diff
@@ -147,55 +140,6 @@ contract Option is IOption, IERC20, European, Ownable {
...
-     * Requires a transaction inclusion proof which is verified by our chain relay.
-     * @dev Can only be called by the parent factory contract.
-     * @param seller Account to exercise against
-     * @param height Bitcoin block height
-     * @param index Bitcoin tx index
-     * @param txid Bitcoin transaction id
-     * @param proof Bitcoin inclusion proof
-     * @param rawtx Bitcoin raw tx
-     **/
-    function executeExercise(
```
Comment (nud3l) [inline contracts/Option.sol:161]: Why did you move the executeExercise out of the option contract?

Comment (gregdhill) [reply contracts/Option.sol:161]: The core logic only pertains to the obligation contract so there is no need to do the forward call in the option contract. For verification we needed `btcHash` and `format` anyway which added additional back and forth complexity.

Comment (nud3l) [reply contracts/Option.sol:161]: As dicussed, the user should have one contract to call when exercising => either the option or the obligation contract.

File: contracts/Obligation.sol [hunk 2]
```diff
@@ -157,19 +163,24 @@ contract Obligation is IObligation, IERC20, European, Ownable, WriterRegistry {
...
      * @param btcHash Bitcoin hash
      * @param format Bitcoin script format
      **/
     function mint(
         address account,
+        address pool,
         uint256 amount,
         bytes20 btcHash,
         Bitcoin.Script format
-    ) external override notExpired onlyOwner {
```
Comment (nud3l) [inline contracts/Obligation.sol:164]: This should remain to be onlyOwner or only be callable internally. Otherwise, anyone can generate obligation tokens that can possibly be sold on Uniswap.

Comment (gregdhill) [reply contracts/Obligation.sol:164]: This is the entrypoint to obligation / option writing which is called by the `OptionLib` for instance. Note that we require the `account` to have sufficient balance in the treasury in order for this call to be valid.

Comment (gregdhill) [reply contracts/Obligation.sol:164]: You can also see that this modifier has been moved to the equivalent method on the `Option` contract since we have reversed the ownership roles.

Comment (nud3l) [reply contracts/Obligation.sol:164]: The function can currently be called directly from any account. If that is done the treasury will try to lock more collateral. If that succeeds because the user has deposited, we generate obligation tokens without also generating option tokens. That is not the desired behaviour.

I agree that you have to kind of force this and on a low level do this, but still, we should not allow it either way. We could e.g. only have this as an internal function.

Comment (gregdhill) [reply contracts/Obligation.sol:164]: So we can only lock 'unlocked' collateral, meaning the user has to atomically supply unlocked collateral then mint (as in Uniswap) - see `lockAndWrite` in `OptionLib.sol`. Once the treasury atomically locks this collateral, preventing additional claims, we then mint the option tokens. Unless we whitelist addresses that can call `mint` here it needs to be external to allow for different 'routers' to do these kinds of atomic operations.

Comment (nud3l) [reply contracts/Obligation.sol:164]: But you could atomically supply, unlock and mint obligation tokens without generating option tokens?

Comment (gregdhill) [reply contracts/Obligation.sol:164]: We call the `mint` function on the option contract as well: https://github.com/interlay/xopts/pull/28/files#diff-c1d6f4d4cca0713bf9049a9a5b64dd8eR197

Comment (nud3l) [reply contracts/Obligation.sol:164]: Ah sorry, my bad! Ok, then this is good to go!

**Review by nud3l — APPROVED** [review]
